### DISCUSSION TITLE
FV convergence

### DISCUSSION THREAD

**Discussion by heinono1** [idaholab/moose#25207] [Q&A General] [answered]
Hi (again). I am checking my FV version of Ferret. I think everything checks out insofar as the output seems to be correct, but the convergence seems to be much poorer compared to the FE version for the same test problem. In particular, the FV version takes many more nonlinear iterations (at smaller timesteps) than the FE version. I wonder if it is generally the case that FV converges more slowly than FE. If that should not be expected, I know I have more work to do ;-). I am using the NewmarkBeta TimeIntegrator both for FE and FV (by far the "best" one for this problem), and asm preconditioner.

**Comment by lindsayad** [comment]
This is not expected. The first thing I would do is check your Jacobian. If those are fine, then we can consider other things. For one, I don't believe the community nor developers have spent a lot of time using time integration different than the default implicit Euler with FV. So I would not be 100% confident there aren't any issues there (good double negative)

**Reply by GiudGiud** [reply]
Agree on both fronts.
- use the Jacobian tester: https://mooseframework.inl.gov/help/development/analyze_jacobian.html
- try implicit euler then bdf2 then Neumark. Do all your equations have a time derivative term?

**Reply by heinono1** [reply]
Hi Alex and Guillaume. Jacobian was the first thing I was thinking of, too, but the FV system uses all AD, so I though I was off the hook on Jacobians....I can of course still use the Jacobian tester to see what comes out of that. 

For the NewmarkBeta time integrator, we (John Mangeri and I) checked different time integrator for the FE version of ferret, and NewmarkBeta was hands down the best one. I have played with some other time integrators for the FV version and NewmarkBeta seems to perform best. But that may be somewhat irrelevant until I have fixed the convergence problem. Not all equations have time derivatives - there are three equations with time derivatives (for the different components of the vector magnetization director) but then also a Poisson equation that is solved simultaneously at each time step with the divergence of the magnetization as a source.

**Reply by lindsayad** [reply]
AD should let you off the hook, but it's always good to have the testing confirmation.

Have you checked that results are the same or are (mesh) converging to the same result between FE and FV? Even if the Jacobian is correct for a given set of residuals, if there are "bugs" in the residuals, then nonlinear convergence may be poor

**Reply by GiudGiud** [reply, marked as answer]
You can accidentally loose the derivative still with AD if you relied on auxkernels or regular non-AD material properties somewhere. It should be fairly rare though, it s definitely easier to get it right with AD. 

So you have a Poisson equation that is FV, and has NO time derivative?
@lindsayad we had an issue for some schemes for the mass equation in NS that had no time derivative. Could be related?

**Reply by heinono1** [reply]
Hi Alex and Guillaume. I started the jacobiandebugger last night. It is still running.

The Poisson Equation has no time derivatives. The nonlinear variables are the three Cartesian components of the magnetization director, and two scalar magnetostatic potentials, one for inside (phi1) the magnetic material, and one (phi2) for outside in the vacuum box. There are Dirichlet BC on phi2 on the boundaries of the vacuum box and at the boundary magnetic-vacuum I enforce continuity (phi1=phi2). I tried to use  FVTwoVarContinuityConstraint to enforce continuity but that did not work for reasons I don't understand: the code would execute but would end up with lambda~0 so and phi2=0 while phi1 was nonzero. So I wrote my own FVInterfaceKernel that does enforce continuity. I have checked that the solution to the Poisson equation for phi1 and phi2 is the same as for the FE version, and the jump discontinuity in derivatives of the magnetostatic potential(s) at the magnetic-vacuum interface is correct. The interface kernel is really simple, just 
```
#include "FVMagneticInterfaceContinuity.h"
#include "MathFVUtils.h"
registerMooseObject("MooseApp", FVMagneticInterfaceContinuity);
InputParameters
FVMagneticInterfaceContinuity::validParams()
{
  InputParameters params = FVInterfaceKernel::validParams();
  params.addClassDescription("Computes the residual of magnetic potential across an interface for "
                             "the finite volume method.");
  return params;
}
FVMagneticInterfaceContinuity::FVMagneticInterfaceContinuity(const InputParameters & params)
  : FVInterfaceKernel(params)
{
}
ADReal
FVMagneticInterfaceContinuity::computeQpResidual()
{
 const auto prefactor = elemIsOne() ? 1 : -1.;
 return  prefactor *( var1().getBoundaryFaceValue(*_face_info, determineState()) -
		      var2().getBoundaryFaceValue(*_face_info, determineState()));
}
```
where var2 and var1 are phi2 and phi1, respectively. It seems to work but maybe it's causing some problems I don't know.

I am running with FDP and lu pc_type. That does change some aspects of the convergence, but not much. The Poisson equation converges more rapidly (residuals of phi1 and phi2 go down to 10^-15 in one iteration) but the residuals of the three magnetic variables do not converge any more rapidly. Pending the completion of the jacobiandebugger it nevertheless seems like the Jacobians are OK as the FDP performs pretty much the same as SMP preconditioning. So this leads me to conclude there is not something not quite right with how I coded up the kernels. I'll keep on digging. 

I really appreciate your rapid responses and your help!

-Olle

**Reply by heinono1** [reply]
It appears that (at least one) problem was how I tried to calculate gradients of the components of the magnetization directors on element faces in an FVFluxKernel. I was using pattern recognition from some NS kernel and had
```
  const auto & mag_x_interface = (has_elem && has_neighbor)
    ? MetaPhysicL::raw_value(_mag_x.gradient(makeCDFace(*_face_info),state))
    : MetaPhysicL::raw_value(_mag_x.gradient(
    makeElemArg(has_elem ? &_face_info->elem() : _face_info->neighborPtr()),state));
  const auto & mag_y_interface = (has_elem && has_neighbor)
      ? MetaPhysicL::raw_value(_mag_y.gradient(makeCDFace(*_face_info),state))
      : MetaPhysicL::raw_value(_mag_y.gradient(
      makeElemArg(has_elem ? &_face_info->elem() : _face_info->neighborPtr()),state));
  const auto & mag_z_interface = (has_elem && has_neighbor)
      ? MetaPhysicL::raw_value(_mag_z.gradient(makeCDFace(*_face_info),state))
      : MetaPhysicL::raw_value(_mag_z.gradient(
      makeElemArg(has_elem ? &_face_info->elem() : _face_info->neighborPtr()),state));
```
When I replace those lines with
```
 const auto & mag_x_interface = _mag_x.gradient(makeCDFace(*_face_info),state);
 const auto & mag_y_interface = _mag_y.gradient(makeCDFace(*_face_info),state);
 const auto & mag_z_interface = _mag_z.gradient(makeCDFace(*_face_info),state);
```
the nonlinear iterations converge as expected. I then tried to add a better interpolation of the gradients:
```
 const auto & mag_x_interface = (has_elem && has_neighbor)
    ?_mag_x.gradient(makeCDFace(*_face_info),state)
    : _mag_x.gradient(
    makeElemArg(has_elem ? &_face_info->elem() : _face_info->neighborPtr()),state);
  const auto & mag_y_interface = (has_elem && has_neighbor)
      ? _mag_y.gradient(makeCDFace(*_face_info),state)
      :_mag_y.gradient(
      makeElemArg(has_elem ? &_face_info->elem() : _face_info->neighborPtr()),state);
  const auto & mag_z_interface = (has_elem && has_neighbor)
      ? _mag_z.gradient(makeCDFace(*_face_info),state)
      : _mag_z.gradient(
      makeElemArg(has_elem ? &_face_info->elem() : _face_info->neighborPtr()),state);
```
which also converges with 2 or 3 nonlinear iterations.

I am running a check now to see if the solutions converge to the same as the FE system that I use as benchmark.

**Reply by lindsayad** [reply]
Oh great news!

**Reply by heinono1** [reply]
Looks like it checks out against the FE version :-)
Thanks for your help!

**Reply by lindsayad** [reply]
🎉
